### ISSUE TITLE
Spades could not be detected

### ISSUE THREAD

**Issue by esud1** [sanger-pathogens/circlator#68]
Hi, 

I am trying to use the CIrclator v1.2.1 in Ubuntu 16.04, but when I am running the progcheck, it says that the program can not detect spades.py. I have put the location of the spades.py in PATH, but to no avail.
I have also attempted to run the install-dependencies.sh script, but it still give me the same result. 

Running "spades.py" and "python2 /path/to/spades.py" on its own was okay.
I have tried both v3.6.0 and 3.6.2 of the Spades assembler.

Here is the outcome of progcheck:

> Circlator version: 1.2.1
> 
> External dependencies:
> bwa	0.7.12	/usr/bin/bwa
> nucmer	3.1	/usr/bin/nucmer
> prodigal	2.6.3	/usr/local/bin/prodigal
> samtools	1.3.1	/usr/local/bin/samtools
> Didn't find spades in path. Looked for:spades.py
> 
> Python version:
> 3.5.2 (default, Jul  5 2016, 12:43:10) 
> [GCC 5.4.0 20160609]
> 
> Python dependencies:
> bio_assembly_refinement	0.5.1	/usr/local/lib/python3.5/dist-packages/bio_assembly_refinement-0.5.1-py3.5.egg/bio_assembly_refinement/__init__.py
> openpyxl	2.4.0-b1	/usr/local/lib/python3.5/dist-packages/openpyxl-2.4.0b1-py3.5.egg/openpyxl/__init__.py
> pyfastaq	3.12.1	/usr/local/lib/python3.5/dist-packages/pyfastaq-3.12.1-py3.5.egg/pyfastaq/__init__.py
> pymummer	0.7.1	/usr/local/lib/python3.5/dist-packages/pymummer-0.7.1-py3.5.egg/pymummer/__init__.py
> pysam	0.8.3	/usr/local/lib/python3.5/dist-packages/pysam-0.8.3-py3.5-linux-x86_64.egg/pysam/__init__.py
> 

And this is outcome of "python2 /path/to/spades.py":

> SPAdes genome assembler v.3.6.0
> 
> Usage: ./build/SPAdes-3.6.0-Linux/bin/spades.py [options] -o <output_dir>
> 
> Basic options:
> -o	<output_dir>	directory to store all the resulting files (required)
> --sc			this flag is required for MDA (single-cell) data
> --iontorrent		this flag is required for IonTorrent data
> --test			runs SPAdes on toy dataset
> -h/--help		prints this usage message
> 
> Input data:
> --12	<filename>	file with interlaced forward and reverse paired-end reads
> -1	<filename>	file with forward paired-end reads
> -2	<filename>	file with reverse paired-end reads
> -s	<filename>	file with unpaired reads
> --pe<#>-12	<filename>	file with interlaced reads for paired-end library number <#> (<#> = 1,2,3,4,5)
> --pe<#>-1	<filename>	file with forward reads for paired-end library number <#> (<#> = 1,2,3,4,5)
> --pe<#>-2	<filename>	file with reverse reads for paired-end library number <#> (<#> = 1,2,3,4,5)
> --pe<#>-s	<filename>	file with unpaired reads for paired-end library number <#> (<#> = 1,2,3,4,5)
> --pe<#>-<or>	orientation of reads for paired-end library number <#> (<#> = 1,2,3,4,5; <or> = fr, rf, ff)
> --s<#>		<filename>	file with unpaired reads for single reads library number <#> (<#> = 1,2,3,4,5)
> --mp<#>-12	<filename>	file with interlaced reads for mate-pair library number <#> (<#> = 1,2,3,4,5)
> --mp<#>-1	<filename>	file with forward reads for mate-pair library number <#> (<#> = 1,2,3,4,5)
> --mp<#>-2	<filename>	file with reverse reads for mate-pair library number <#> (<#> = 1,2,3,4,5)
> --mp<#>-s	<filename>	file with unpaired reads for mate-pair library number <#> (<#> = 1,2,3,4,5)
> --mp<#>-<or>	orientation of reads for mate-pair library number <#> (<#> = 1,2,3,4,5; <or> = fr, rf, ff)
> --hqmp<#>-12	<filename>	file with interlaced reads for high-quality mate-pair library number <#> (<#> = 1,2,3,4,5)
> --hqmp<#>-1	<filename>	file with forward reads for high-quality mate-pair library number <#> (<#> = 1,2,3,4,5)
> --hqmp<#>-2	<filename>	file with reverse reads for high-quality mate-pair library number <#> (<#> = 1,2,3,4,5)
> --hqmp<#>-s	<filename>	file with unpaired reads for high-quality mate-pair library number <#> (<#> = 1,2,3,4,5)
> --hqmp<#>-<or>	orientation of reads for high-quality mate-pair library number <#> (<#> = 1,2,3,4,5; <or> = fr, rf, ff)
> --nxmate<#>-1	<filename>	file with forward reads for Lucigen NxMate library number <#> (<#> = 1,2,3,4,5)
> --nxmate<#>-2	<filename>	file with reverse reads for Lucigen NxMate library number <#> (<#> = 1,2,3,4,5)
> --sanger	<filename>	file with Sanger reads
> --pacbio	<filename>	file with PacBio reads
> --nanopore	<filename>	file with Nanopore reads
> --trusted-contigs	<filename>	file with trusted contigs
> --untrusted-contigs	<filename>	file with untrusted contigs
> 
> Pipeline options:
> --only-error-correction	runs only read error correction (without assembling)
> --only-assembler	runs only assembling (without read error correction)
> --careful		tries to reduce number of mismatches and short indels
> --continue		continue run from the last available check-point
> --restart-from	<cp>	restart run with updated options and from the specified check-point ('ec', 'as', 'k<int>', 'mc')
> --disable-gzip-output	forces error correction not to compress the corrected reads
> --disable-rr		disables repeat resolution stage of assembling
> 
> Advanced options:
> --dataset	<filename>	file with dataset description in YAML format
> -t/--threads	<int>		number of threads
> 				[default: 16]
> -m/--memory	<int>		RAM limit for SPAdes in Gb (terminates if exceeded)
> 				[default: 250]
> --tmp-dir	<dirname>	directory for temporary files
> 				[default: <output_dir>/tmp]
> -k		<int,int,...>	comma-separated list of k-mer sizes (must be odd and
> 				less than 128) [default: 'auto']
> --cov-cutoff	<float>		coverage cutoff value (a positive float number, or 'auto', or 'off') [default: 'off']
> --phred-offset	<33 or 64>	PHRED quality offset in the input reads (33 or 64)
> 				[default: auto-detect]
> 

Can you please advice on this matter? 
Thanks

**Comment by martinghunt** [comment]
Doing this might fix it:

   export CIRCLATOR_SPADES=/your/path/to/spades.py

But could you also let me know what you get if you run this:

    python3 -c 'import shutil; print(shutil.which("spades.py"))'

as it would help me understand what is happening.

Thanks.

**Comment by esud1** [comment]
Hi Martin, 

I tried the export command, but the progcheck still gives me the same outcome. 

Here is the result of the command you suggested: 

> /home/ubuntu/bioinformatics/SPAdes-3.6.0-Linux/bin/spades.py

It prints the location of the spades.py

Thanks

**Comment by martinghunt** [comment]
Thanks for the info. I'm working on a new release of circlator at the moment. It's hard to know for sure what's going wrong, but I think I can fix this. I'll get back to you in a few days when the release is ready...

**Comment by esud1** [comment]
Thanks! 
Looking forward for the new release

**Comment by martinghunt** [comment]
Before I make the release, could you test it out please by cloning/downloading the master branch? I'm hoping the latest changes will have sorted the spades issues.
Thanks!

**Comment by esud1** [comment]
Hi Martin, 

Sorry for the late reply. 

I tried the new circlator, and downloaded the SPAdes v.3.7.1
The spades itself has no issues, I can call it by typing spades.py, but the progcheck still have the same problem - "Didn't find spades in path. Looked for:spades.py"

So I tried to copy the whole files in SPAdes bin to /usr/bin, and now the progcheck gave me this message: 
"Found spades but couldn't get version."

Thanks

**Comment by martinghunt** [comment]
How about trying this in an empty directory?

```
wget http://spades.bioinf.spbau.ru/release3.7.1/SPAdes-3.7.1-Linux.tar.gz
tar zxf SPAdes-3.7.1-Linux.tar.gz
export CIRCLATOR_SPADES=$PWD/SPAdes-3.7.1-Linux/bin/spades.py
circlator progcheck
```

... please don't change directories or move anything, just run those 4 commands. I guess you might have to give the full path to circlator if it's not installed globally. Thanks.

**Comment by esud1** [comment]
Hi, 

Here is the result that I get from running only those 4 commands in a new and empty directory: 

> Circlator version: 1.2.1
> 
> External dependencies:
> bwa	0.7.12	/usr/bin/bwa
> nucmer	3.1	/usr/bin/nucmer
> prodigal	2.6.3	/usr/local/bin/prodigal
> samtools	1.3.1	/usr/local/bin/samtools
> Found spades but couldn't get version.
> 
> Python version:
> 3.5.2 (default, Jul  5 2016, 12:43:10) 
> [GCC 5.4.0 20160609]
> 
> Python dependencies:
> openpyxl	2.4.0-b1	/usr/local/lib/python3.5/dist-packages/openpyxl-2.4.0b1-py3.5.egg/openpyxl/__init__.py
> pyfastaq	3.12.1	/usr/local/lib/python3.5/dist-packages/pyfastaq-3.12.1-py3.5.egg/pyfastaq/__init__.py
> pymummer	0.7.1	/usr/local/lib/python3.5/dist-packages/pymummer-0.7.1-py3.5.egg/pymummer/__init__.py
> pysam	0.8.3	/usr/local/lib/python3.5/dist-packages/pysam-0.8.3-py3.5-linux-x86_64.egg/pysam/__init__.py

**Comment by martinghunt** [comment]
Thanks for your patience :)  Could you get the latest code from the master branch? I just added a --debug option to progcheck that is very verbose. Should tell us what's going wrong:

    circlator progcheck --debug

**Comment by esud1** [comment]
Hi, 

Here is the output: 

> Circlator version: 1.2.1
> 
> External dependencies:
> __________ checking bwa ____________
> bwa - not using environment variable
> bwa - checking which(bwa)
>    ... got: "/usr/bin/bwa"
> bwa - checking version ...
> Running this command to get version: bwa 
> __________ (begin output from bwa )___________
> 
> Program: bwa (alignment via Burrows-Wheeler transformation)
> Version: 0.7.12-r1039
> Contact: Heng Li <lh3@sanger.ac.uk>
> 
> Usage:   bwa <command> [options]
> 
> Command: index         index sequences in the FASTA format
>          mem           BWA-MEM algorithm
>          fastmap       identify super-maximal exact matches
>          pemerge       merge overlapping paired ends (EXPERIMENTAL)
>          aln           gapped/ungapped alignment
>          samse         generate alignment (single ended)
>          sampe         generate alignment (paired ended)
>          bwasw         BWA-SW for long queries
> 
>          shm           manage indices in shared memory
>          fa2pac        convert FASTA to PAC format
>          pac2bwt       generate BWT from PAC
>          pac2bwtgen    alternative algorithm for generating BWT
>          bwtupdate     update .bwt to the new format
>          bwt2sa        generate SA from BWT and Occ
> 
> Note: To use BWA, you need to first index the genome with `bwa index'.
>       There are three alignment algorithms in BWA: `mem', `bwasw', and
>       `aln/samse/sampe'. If you are not sure which to use, try `bwa mem'
>       first. Please `man ./bwa.1' for the manual.
> 
> __________ (end of output from bwa )___________
> Looking far a match to the regex "^Version: ([0-9\.]+)" in the above output
> Match to this line: Version: 0.7.12-r1039
> Got version: 0.7.12
> bwa	0.7.12	/usr/bin/bwa
> __________ checking nucmer ____________
> nucmer - not using environment variable
> nucmer - checking which(nucmer)
>    ... got: "/usr/bin/nucmer"
> nucmer - checking version ...
> Running this command to get version: nucmer --version
> __________ (begin output from nucmer --version)___________
> nucmer 
> NUCmer (NUCleotide MUMmer) version 3.1
>     
> __________ (end of output from nucmer --version)___________
> Looking far a match to the regex "^NUCmer \(NUCleotide MUMmer\) version ([0-9\.]+)" in the above output
> Match to this line: NUCmer (NUCleotide MUMmer) version 3.1
> Got version: 3.1
> nucmer	3.1	/usr/bin/nucmer
> __________ checking prodigal ____________
> prodigal - not using environment variable
> prodigal - checking which(prodigal)
>    ... got: "/usr/local/bin/prodigal"
> prodigal - checking version ...
> Running this command to get version: prodigal -v
> __________ (begin output from prodigal -v)___________
> 
> Prodigal V2.6.3: February, 2016
> 
> __________ (end of output from prodigal -v)___________
> Looking far a match to the regex "^Prodigal V([0-9\.]+):" in the above output
> Match to this line: Prodigal V2.6.3: February, 2016
> Got version: 2.6.3
> prodigal	2.6.3	/usr/local/bin/prodigal
> __________ checking samtools ____________
> samtools - not using environment variable
> samtools - checking which(samtools)
>    ... got: "/usr/local/bin/samtools"
> samtools - checking version ...
> Running this command to get version: samtools 
> __________ (begin output from samtools )___________
> 
> Program: samtools (Tools for alignments in the SAM format)
> Version: 1.3.1 (using htslib 1.3.1)
> 
> Usage:   samtools <command> [options]
> 
> Commands:
>   -- Indexing
>      dict           create a sequence dictionary file
>      faidx          index/extract FASTA
>      index          index alignment
> 
>   -- Editing
>      calmd          recalculate MD/NM tags and '=' bases
>      fixmate        fix mate information
>      reheader       replace BAM header
>      rmdup          remove PCR duplicates
>      targetcut      cut fosmid regions (for fosmid pool only)
>      addreplacerg   adds or replaces RG tags
> 
>   -- File operations
>      collate        shuffle and group alignments by name
>      cat            concatenate BAMs
>      merge          merge sorted alignments
>      mpileup        multi-way pileup
>      sort           sort alignment file
>      split          splits a file by read group
>      quickcheck     quickly check if SAM/BAM/CRAM file appears intact
>      fastq          converts a BAM to a FASTQ
>      fasta          converts a BAM to a FASTA
> 
>   -- Statistics
>      bedcov         read depth per BED region
>      depth          compute the depth
>      flagstat       simple stats
>      idxstats       BAM index stats
>      phase          phase heterozygotes
>      stats          generate stats (former bamcheck)
> 
>   -- Viewing
>      flags          explain BAM flags
>      tview          text alignment viewer
>      view           SAM<->BAM<->CRAM conversion
>      depad          convert padded BAM to unpadded BAM
> 
> __________ (end of output from samtools )___________
> Looking far a match to the regex "^Version: ([0-9\.]+)" in the above output
> Match to this line: Version: 1.3.1 (using htslib 1.3.1)
> Got version: 1.3.1
> samtools	1.3.1	/usr/local/bin/samtools
> __________ checking spades ____________
> spades.py - not using environment variable
> spades.py - checking which(spades.py)
>    ... got: "None"
> spades.py - checking version ...
>  ... not in path so cannot get version
> Didn't find spades in path. Looked for:spades.py
> 
> Python version:
> 3.5.2 (default, Jul  5 2016, 12:43:10) 
> [GCC 5.4.0 20160609]
> 
> Python dependencies:
> openpyxl	2.4.0-b1	/usr/local/lib/python3.5/dist-packages/openpyxl-2.4.0b1-py3.5.egg/openpyxl/__init__.py
> pyfastaq	3.12.1	/usr/local/lib/python3.5/dist-packages/pyfastaq-3.12.1-py3.5.egg/pyfastaq/__init__.py
> pymummer	0.7.1	/usr/local/lib/python3.5/dist-packages/pymummer-0.7.1-py3.5.egg/pymummer/__init__.py
> pysam	0.8.3	/usr/local/lib/python3.5/dist-packages/pysam-0.8.3-py3.5-linux-x86_64.egg/pysam/__init__.py
>

**Comment by martinghunt** [comment]
So it looks like spades.py is not in your $PATH. Can you try this again:

    wget http://spades.bioinf.spbau.ru/release3.7.1/SPAdes-3.7.1-Linux.tar.gz
    tar zxf SPAdes-3.7.1-Linux.tar.gz
    export CIRCLATOR_SPADES=$PWD/SPAdes-3.7.1-Linux/bin/spades.py
    circlator progcheck

**Comment by esud1** [comment]
Hi Martin, 

I have repeated the steps above, and the spades.py is in my $PATH, but it still gives me message that spades.py cannot be found.
A double check of `printenv` shows the presence of spades.py in the $PATH. 
Here is the excerpt of `printenv`: 

> PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ubuntu/bioinformatics/Reapr_1.0.18/:/home/ubuntu/abc/SPAdes-3.7.1-Linux/bin/spades.py
> CIRCLATOR_SPADES=/home/ubuntu/abc/SPAdes-3.7.1-Linux/bin/spades.py
>

**Comment by martinghunt** [comment]
Unix needs directories in $PATH. Try removing the "spades.py" so your path has this in it:

    /home/ubuntu/bioinformatics/Reapr_1.0.18/:/home/ubuntu/abc/SPAdes-3.7.1-Linux/bin/

and then run progcheck with the --debug option.

**Comment by GermainChev** [comment]
Hello 

I have the same kind of problem except that for me progcheck find spades but he couldn't get the version ...

circlator progcheck
Circlator version: 1.2.1

External dependencies:
bwa	0.7.12	/Users/germain/Desktop/Soft.dir/circlator-1.2.1/build/bwa-0.7.12/bwa
nucmer	3.1	/Users/germain/Desktop/Soft.dir/circlator-1.2.1/build/MUMmer3.23/nucmer
prodigal	2.6.3	/Users/germain/Desktop/Soft.dir/circlator-1.2.1/build/prodigal-2.6.2/prodigal
samtools	1.3	/Users/germain/Desktop/Soft.dir/circlator-1.2.1/build/samtools-1.3/samtools
Found spades but couldn't get version.

Python version:
3.5.1 (v3.5.1:37a07cee5969, Dec  5 2015, 21:12:44) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]

Python dependencies:
bio_assembly_refinement	0.5.1	/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/bio_assembly_refinement-0.5.1-py3.5.egg/bio_assembly_refinement/__init__.py
openpyxl	2.4.0-b1	/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/openpyxl-2.4.0b1-py3.5.egg/openpyxl/__init__.py
pyfastaq	3.13.0	/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pyfastaq-3.13.0-py3.5.egg/pyfastaq/__init__.py
pymummer	0.7.1	/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pymummer-0.7.1-py3.5.egg/pymummer/__init__.py
pysam	0.8.3	/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pysam-0.8.3-py3.5-macosx-10.6-intel.egg/pysam/__init__.py

Thanks for your help !!

**Comment by martinghunt** [comment]
Hi @GermainChev 

I just made a new release, which I think will fix the problem as you're using a mac (probably the same as issue #67). If it doesn't, could you let me know the output of running

    circlator progcheck --debug

Thanks.

**Comment by GermainChev** [comment]
Thanks martinghunt !!
I installed the new release 
I used install_dependencies.sh
So I still have the same pb ...
Moreover I got the same error with prodigal 
Then I replaced prodigal-2.6.2 by prodigal-2.6 and it fix it (he could find the version)
But regarding spades he's still not able to find the version (I also tried SPAdes-3.5.0)
Finally parameter --debug  is a unrecognized arguments ...

Thanks!!

**Comment by martinghunt** [comment]
Are you sure you have the new release 1.3.0? I just downloaded it to double-check and it has the --debug option...

```
usage: circlator progcheck

Checks all dependencies are found and are correct versions

optional arguments:
  -h, --help  show this help message and exit
  --debug     Debug mode with very verbose output
```

Maybe your install is still picking up the previous version?

**Comment by GermainChev** [comment]
Sorry I forgot to run python3 setup.py install ....

Now It detect everything !! And it works !!

Thanks again !!

**Comment by esud1** [comment]
Hi Martin, 

I have just tried the newly released v1.3
And now it works for me as well!
Thanks a lot for the help :)
